agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -2585,95 +2616,118 @@ bool backhaul_manager::handle_ap_metrics_query(ieee1905_1::CmduMessageRx &cmdu_r
             LOG(ERROR) << "Failed to get bssid " << bssid_idx << " from AP_METRICS_QUERY";
             return false;
         }
-        bssid.push_back(std::get<1>(bssid_tuple));
+
+        m_expected_ap_metrics_response.add_expected_bssid(std::get<1>(bssid_tuple));
+
         LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
                    << "  bssid " << std::get<1>(bssid_tuple);
     }
 
-    if (!send_slave_ap_metric_query_message(mid, bssid)) {
-        LOG(ERROR) << "Failed to forward AP_METRICS_RESPONSE to the son_slave_thread";
+    LOG(DEBUG) << "Forwarding AP_METRICS_QUERY_MESSAGE to all slaves, mid=" << std::hex << int(mid);
+    auto header = message_com::get_uds_header(cmdu_rx);
+    if (!header) {
+        LOG(ERROR) << "Failed to extract header from rx message";
         return false;
     }
 
+    uint16_t length = header->length;
+    cmdu_rx.swap(); // swap back before forwarding
+    for (const auto &soc : slaves_sockets) {
+        if (!message_com::forward_cmdu_to_uds(soc->slave, cmdu_rx, length)) {
+            LOG(ERROR) << "Failed forwarding AP_METRICS_QUERY_MESSAGE message to slave: "
+                       << tlvf::mac_to_string(soc->radio_mac);
+        }
+    }
+
     return true;
 }
 
 bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageRx &cmdu_rx,
                                                         const std::string &src_mac)
 {
     auto mid = cmdu_rx.getMessageId();
-    LOG(DEBUG) << "Received AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
 
-    /**
-     * If AP Metrics Response message does not correspond to a previously received and forwarded
-     * AP Metrics Query message (which we know because message id is not set), then forward message
-     * to controller.
-     * This might happen when channel utilization value has crossed configured threshold or when
-     * periodic metrics reporting interval has elapsed.
-     */
-    if (0 == mid) {
-        uint16_t length = message_com::get_uds_header(cmdu_rx)->length;
-        cmdu_rx.swap(); //swap back before forwarding
-        return send_cmdu_to_bus(cmdu_rx, controller_bridge_mac, bridge_info.mac, length);
+    if (m_expected_ap_metrics_response.get_mid() != mid) {
```

Comment:
We had a long discussion regarding how to handle AP Metrics Responses correctly for all cases: specific request, repeated requests and spontaneous AP metrics response - taking into account that mid=0 is legitimate for all cases.
There was no `correct and simple` solution.
Options were either simple or correct, but not both.
Anyhow I moved the `forward unconditionally on mid=0` to be the first check.

Reply:
You shall not unconditionally forward responses from slaves to controller if mid is equal to 0.
If you do so, responses to the query in case of periodic metrics reporting are not gathered together in a single response.
The proof is that tests are now failing.

Please check it, but I think that restoring the code as it was when I wrote my comment and calling `m_expected_ap_metrics_response.reset_to_new_mid(0);` if `m_expected_ap_metrics_response.is_expected_bssid_empty()`, right before sending AP Metrics Response message will do the job. Well, at least if the mid in AP Metrics Query message was not also 0, but this will happen once in a blue moon only.

> There was no correct and simple solution.

My suggestion is yet another another patch to the list of patches used to implement this flow. The problem we are trying to solve is or resembles a lot to a well known problem called *stack ripping*. I've been thinking on how to correctly implement this flow by addressing the root cause of the issue and I think that I've come up with a solution: using one new sequence byte in the UDS header called `rid` (for "request ID") would allow us to correctly match queries and responses. The trick is that, unlike the mid value, the rid would not be allowed to be 0, as this value would be reserved for responses without a query (for the channel utilization use case). 

Maybe it's too late for this PR, but I had to write this just in case it can be used to implement a similar flow in the future.